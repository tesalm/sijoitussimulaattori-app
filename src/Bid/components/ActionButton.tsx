import React from 'react';

import { Text, View } from 'react-native';
import { WizardFormColors } from '../../App/colors';
import { t } from '../../assets/i18n';
import Icon from '../../general/icon';
import { actionButtons } from '../styles';

export interface ActionButtonProps {
  action: string;
  active: boolean;
}

export const ActionButton = (props: ActionButtonProps): JSX.Element => {
  return (
    <View style={actionButtons.buttonWithText}>
      <Icon
        iconName={props.action}
        iconHeight={24}
        iconWidth={24}
        tintColor={
          props.active
            ? WizardFormColors.buttonsActive
            : WizardFormColors.buttonsUnactive
        }
      />
      <Text
        style={
          props.active
            ? actionButtons.buttonActive
            : actionButtons.buttonUnactive
        }
      >
        {props.action === 'buy' && t('BidPage.Buy')}
        {props.action === 'sell' && t('BidPage.Sell')}
      </Text>
    </View>
  );
};
