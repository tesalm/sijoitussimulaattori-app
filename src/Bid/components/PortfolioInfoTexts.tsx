import React from 'react';
import { Text, View } from 'react-native';

import { t } from '../../assets/i18n';
import { SinglePortfolio } from '../../PortfolioList/reducer';
import { formatCurrency } from '../../util/general';
import { bidStyles } from '../styles';

export interface PortfolioInfoTextsProps {
  currentStockSymbol?: string;
  portfolioList?: SinglePortfolio[];
  selectedPortfolio?: string;
  stockName: string;
}

// Returns info of the selected portfolio to the bid-screen
export const PortfolioInfoTexts = (
  props: PortfolioInfoTextsProps
): JSX.Element => {
  // Returns selected portfolio as SinglePortfolio.
  function getCurrentPortfolio() {
    if (props.portfolioList) {
      return props.portfolioList.find((portfolio) => {
        return portfolio.name === props.selectedPortfolio;
      });
    }
  }

  // Get the sum of the current stocks in selected portfolio. If there is no stocks, returns 0.
  function getSumOfStockFromPortfolio() {
    const currentPortfolio = getCurrentPortfolio();
    if (
      currentPortfolio &&
      currentPortfolio.portfolioInfo &&
      currentPortfolio.portfolioInfo.portfolio
    ) {
      if (!currentPortfolio.portfolioInfo.portfolio.stocks) {
        return 0;
      } else {
        const portfolioStocks = currentPortfolio.portfolioInfo.portfolio.stocks.find(
          (stock) => {
            return stock.uid === props.currentStockSymbol;
          }
        );
        if (portfolioStocks) {
          return portfolioStocks.amount;
        }
      }
    }
    return 0;
  }

  // TODO: Take away default currency.
  // Returns the total market value of selected portfolio.
  function getPortfolioValue() {
    const currentPortfolio = getCurrentPortfolio();
    if (
      currentPortfolio &&
      currentPortfolio.portfolioInfo &&
      currentPortfolio.portfolioInfo.portfolio
    ) {
      return formatCurrency(
        currentPortfolio.portfolioInfo.portfolio.totalMarketValue,
        '$'
      );
    } else {
      return '0$';
    }
  }

  return (
    <View>
      <Text style={bidStyles.infoText}>
        {t('SumUpPage.YouOwn')}{' '}
        <Text style={bidStyles.infoTextHighlight}>
          {getSumOfStockFromPortfolio()}
        </Text>{' '}
        {t('SumUpPage.StocksOf')} {props.stockName}.
      </Text>
      <Text style={bidStyles.infoText}>
        {t('SumUpPage.YourPortfoliosTotVal')}{' '}
        <Text style={bidStyles.infoTextHighlight}>{getPortfolioValue()}</Text>
      </Text>
    </View>
  );
};
