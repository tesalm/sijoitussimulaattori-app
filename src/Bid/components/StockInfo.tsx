import React from 'react';
import { Text, View } from 'react-native';

import { t } from '../../assets/i18n';
import { getLocale } from '../../util/general';
import { formatRevenue } from '../../util/stock';
import { stockinfo } from '../styles';

export interface StockInfoProps {
  name: string;
  revenue: number;
  updated: Date;
}

export const StockInfo = (props: StockInfoProps): JSX.Element => {
  return (
    <View style={stockinfo.container}>
      <View>
        <Text style={stockinfo.stockText}>{props.name}</Text>
      </View>
      <View style={stockinfo.stockinfoMiddle}>
        <View style={stockinfo.stockinfoMiddleContent}>
          <Text style={stockinfo.valueHeaderSmall}>
            {t('StockPage.RevenueText')}
          </Text>
          <Text style={stockinfo.valueSmall}>
            {formatRevenue(props.revenue)}
          </Text>
        </View>
        <View style={stockinfo.stockinfoMiddleContent}>
          {/* TODO: Here comes the graph */}
        </View>
      </View>
      <View>
        <Text style={stockinfo.updatedText}>
          {t('StockPage.Updated')}: {props.updated.toLocaleString(getLocale())}
        </Text>
      </View>
    </View>
  );
};
