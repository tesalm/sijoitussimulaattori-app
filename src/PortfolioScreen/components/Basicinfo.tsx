import React from 'react';
import { ActivityIndicator, Text, View } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { t } from '../../assets/i18n';
import { RootState } from '../../redux/reducers';
import { Portfolio } from '../reducers';
import { stockStyles } from '../styles';

export interface BasicinfoProps {
  portfolio?: Portfolio;
  loading?: boolean;
  error?: Error;
}

export class Basicinfo extends React.Component<BasicinfoProps> {
  constructor(props: BasicinfoProps) {
    super(props);
  }
  revenueColor = (revenue: number): typeof stockStyles.revenueValueGreen => {
    return revenue >= 0
      ? stockStyles.revenueValueGreen
      : stockStyles.revenueValueRed;
  };

  formatValue = (value: number, currency: string): string => {
    if (currency == 'USD') {
      return value + ' $';
    } else if (currency == 'EUR') {
      return value + ' €';
    }
    return value + ' $';
  };

  render() {
    const { portfolio, loading, error } = this.props;
    if (loading == undefined && portfolio == undefined && error == undefined) {
      return (
        <View>
          <Text>vituiks meni</Text>
        </View>
      );
    }

    if (loading) {
      return (
        <View style={stockStyles.loading}>
          <ActivityIndicator size="large" />
        </View>
      );
    } else {
      if (portfolio == undefined) {
        //TODO: Format the error message to user
        if (error) {
          return <Text>Error! {error.message} </Text>;
        } else {
          return <Text>Error! asöfeh {portfolio} </Text>;
        }
      }
    }

    return (
      <View>
        <View style={stockStyles.basicinfo}>
          <Text style={stockStyles.titleStyle}>{t('PortfolioPage.Title')}</Text>
        </View>
        <View style={stockStyles.basicinfo}>
          <View style={stockStyles.basicinfoSmallerComp}>
            <Text style={stockStyles.valueHeader}>
              {t('PortfolioPage.MarketValue')}
            </Text>
            <Text style={stockStyles.value}>{portfolio.marketValue}</Text>
            <Text style={stockStyles.valueHeader}>
              {t('PortfolioPage.Cash')}
            </Text>
            <Text style={stockStyles.value}>{portfolio.cash} </Text>
          </View>
          <View style={stockStyles.basicinfoSmallerComp}>
            <Text style={stockStyles.valueHeaderMiddle}>
              {t('PortfolioPage.TotalValue')}
            </Text>
            <Text style={stockStyles.valueMiddle}>
              {portfolio.marketValue + portfolio.cash}
            </Text>
          </View>

          <View style={stockStyles.basicinfoMidComp}>
            <Text style={stockStyles.valueHeaderRightSide}>
              {t('PortfolioPage.Revenue')}
            </Text>
            <Text style={this.revenueColor(4.52)}>{portfolio.revenue}</Text>
          </View>
        </View>
      </View>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  portfolio: state.singlePortfolio.portfolio,
  loading: state.singlePortfolio.loading,
  error: state.singlePortfolio.error,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators({}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Basicinfo);
