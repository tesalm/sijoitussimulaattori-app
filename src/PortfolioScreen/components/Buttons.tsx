import React from 'react';
import { Text, TouchableHighlight, View } from 'react-native';

import { generalStyles } from '../../App/generalStyles';
import { t } from '../../assets/i18n';
import Icon, { IconNames } from '../../general/icon';

export const EventsTransactions = (): JSX.Element => {
  return (
    // TODO: Make onPress for these items that go to buy or sell view.
    <TouchableHighlight style={generalStyles.button}>
      <View style={generalStyles.buttonView}>
        <View style={generalStyles.buttonLogoView}>
          <Icon iconName={IconNames.manage} iconHeight={24} iconWidth={24} />
          <Text style={generalStyles.buttonTitle}>
            {t('PortfolioPage.Events')}
          </Text>
        </View>
        <Icon iconName={IconNames.open} iconHeight={24} iconWidth={24} />
      </View>
    </TouchableHighlight>
  );
};

export const Manage = (): JSX.Element => {
  return (
    // TODO: Make onPress for these items that go to buy or sell view.
    <TouchableHighlight style={generalStyles.button}>
      <View style={generalStyles.buttonView}>
        <View style={generalStyles.buttonLogoView}>
          <Icon iconName={IconNames.manage} iconHeight={24} iconWidth={24} />
          <Text style={generalStyles.buttonTitle}>
            {t('PortfolioPage.Manage')}
          </Text>
        </View>
        <Icon iconName={IconNames.open} iconHeight={24} iconWidth={24} />
      </View>
    </TouchableHighlight>
  );
};
