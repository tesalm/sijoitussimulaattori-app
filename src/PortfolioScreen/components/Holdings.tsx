import React from 'react';
import { ActivityIndicator, Text, View } from 'react-native';
import Accordion from 'react-native-collapsible/Accordion';

import { t } from '../../assets/i18n';
import Icon from '../../general/icon';
import { Stock } from '../../MarketScreen/reducer';
import {
  calculateTotalMarketValue,
  calculateTotalRevenue,
  calculateTotalRevenueProcent,
  countRevenuePercentage,
  formatCurrency,
  formatRevenue,
  formatRevenueCurrency,
  revenueColor,
  valueColor,
} from '../../util/stock';
import { Portfolio, PortfolioStock } from '../reducers';
import { portfolioStyles } from '../styles';

interface HoldingsProps {
  portfolio?: Portfolio;
  loading?: boolean;
  error?: Error;
  stocks?: Array<Stock>;
}

interface HoldingsState {
  activeSections: Array<number>;
}

export class Holdings extends React.Component<HoldingsProps, HoldingsState> {
  constructor(props: HoldingsProps, state: HoldingsState) {
    super(props);
    this.state = { activeSections: [] };
  }

  renderHeader = (
    section: PortfolioStock,
    index: number,
    isActive: boolean
  ) => {
    const iconName = isActive ? 'arrowUp' : 'arrowDown';

    return (
      <View style={portfolioStyles.holdingsSubLogoView}>
        <View style={portfolioStyles.holdingsSubTitleView}>
          <Text style={portfolioStyles.holdingsSubTitle}>{section.name}</Text>
        </View>
        <View style={portfolioStyles.holdingsArrowView}>
          <Icon iconName={iconName} iconHeight={24} iconWidth={24} />
        </View>
      </View>
    );
  };

  renderContent = (section: PortfolioStock) => {
    const { error, stocks } = this.props;
    if (stocks == undefined) {
      let errorMessage;
      if (error) {
        errorMessage = error.message + ' ';
      }
      // TODO: Muokkaa error-teksti käyttäjälle.
      return <Text>Error! {errorMessage} </Text>;
    }

    //find right stock from the array.
    const rightStock = stocks.find((stock) => stock.symbol == section.symbol);
    if (rightStock == undefined) {
      let errorMessage;
      if (error) {
        errorMessage = error.message + ' ';
      }
      // TODO: Muokkaa error-teksti käyttäjälle.
      return <Text>Error! {errorMessage} </Text>;
    }
    if (
      rightStock.stockInfo.intraday === undefined ||
      rightStock.stockInfo.stockMetadata === undefined ||
      rightStock.stockInfo.historyData == undefined
    ) {
      let errorMessage;
      if (rightStock.stockInfo.metaError) {
        errorMessage = rightStock.stockInfo.metaError.message + ' ';
      }
      if (rightStock.stockInfo.intraError) {
        errorMessage = errorMessage + rightStock.stockInfo.intraError.message;
      }
      if (rightStock.stockInfo.historyError) {
        errorMessage = errorMessage + rightStock.stockInfo.historyError.message;
      }
      // TODO: Muokkaa error-teksti käyttäjälle.
      return <Text>Error! {errorMessage} </Text>;
    }

    //calculate all necessary values needed.
    const currency = rightStock.stockInfo.stockMetadata.currency;
    const close = rightStock.stockInfo.intraday.intradayQuote[0].close;
    const totalRevenueProcent = calculateTotalRevenueProcent(
      rightStock,
      section
    );
    const totalMarketValue = calculateTotalMarketValue(rightStock, section);
    const totalRevenue = calculateTotalRevenue(rightStock, section);
    const revenueProcent = countRevenuePercentage(rightStock);

    return (
      <View>
        <View style={portfolioStyles.holdingsContainer}>
          <View style={portfolioStyles.portfolioInfoSmallerComp}>
            <Text style={portfolioStyles.valueHeader}>
              {t('PortfolioPage.StocksOwned')}
            </Text>
            <Text style={portfolioStyles.value}>{section.amount}</Text>
            <Text style={portfolioStyles.valueHeader}>
              {t('PortfolioPage.MarketValue')}
            </Text>
            <Text style={portfolioStyles.value}>
              {formatCurrency(close, currency)}{' '}
            </Text>
            <Text style={portfolioStyles.valueHeader}>
              {t('PortfolioPage.TotalRevenueProcent')}
            </Text>
            <Text style={valueColor(totalRevenueProcent)}>
              {formatRevenue(totalRevenueProcent)}{' '}
            </Text>
          </View>
          <View style={portfolioStyles.portfolioInfoSmallerComp}>
            <Text style={portfolioStyles.valueHeader}>
              {t('PortfolioPage.Acquisition')}
            </Text>
            <Text style={portfolioStyles.value}>
              {formatCurrency(section.avgPrice, currency)}
            </Text>
            <Text style={portfolioStyles.valueHeader}>
              {t('PortfolioPage.TotalMarketValue')}
            </Text>
            <Text style={portfolioStyles.value}>
              {formatCurrency(totalMarketValue, currency)}{' '}
            </Text>
            <Text style={portfolioStyles.valueHeader}>
              {t('PortfolioPage.TotalRevenue')}
            </Text>
            <Text style={valueColor(totalRevenue)}>
              {formatRevenueCurrency(totalRevenue, currency)}{' '}
            </Text>
          </View>

          <View style={portfolioStyles.portfolioInfoSmallerComp}>
            <Text style={portfolioStyles.valueHeaderRightSideHoldings}>
              {t('PortfolioPage.Revenue')}
            </Text>
            <Text style={revenueColor(revenueProcent)}>
              {formatRevenue(revenueProcent)}
            </Text>
          </View>
        </View>
      </View>
    );
  };

  updateSections = (activeSections: number[]) => {
    this.setState({ activeSections });
  };

  render() {
    const { portfolio, error, loading, stocks } = this.props;
    if (loading) {
      return (
        <View style={portfolioStyles.loading}>
          <ActivityIndicator size="large" />
        </View>
      );
    } else {
      if (
        portfolio === undefined ||
        stocks == undefined ||
        portfolio == undefined
      ) {
        let errorMessage;
        if (error) {
          errorMessage = error.message + ' ';
        }
        // TODO: Muokkaa error-teksti käyttäjälle.
        return <Text>Error! {errorMessage} </Text>;
      }
    }

    return (
      <View>
        <View style={portfolioStyles.holdingsLogoView}>
          <Icon iconName={'holdings'} iconHeight={24} iconWidth={24} />
          <Text style={portfolioStyles.titleText}>
            {t('PortfolioPage.Holdings')}
          </Text>
        </View>
        <Accordion
          expandMultiple={true}
          sections={portfolio.stocks}
          activeSections={this.state.activeSections}
          renderHeader={this.renderHeader}
          renderContent={this.renderContent}
          onChange={this.updateSections}
          sectionContainerStyle={portfolioStyles.accordionContainer}
        />
      </View>
    );
  }
}
