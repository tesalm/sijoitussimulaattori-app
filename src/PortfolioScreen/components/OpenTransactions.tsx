import React from 'react';
import { ActivityIndicator, Alert, Text, TouchableOpacity, View } from 'react-native';
import Collapsible from 'react-native-collapsible';
import { scale, verticalScale } from 'react-native-size-matters';

import { t } from '../../assets/i18n';
import Icon from '../../general/icon';
import { cancelTransaction, getTransactions } from '../../PortfolioList/actions';
import { Transaction } from '../../PortfolioList/reducer';
import { portfolioStyles } from '../styles';

export interface TransactionProps {
  transactions: Array<Transaction>;
  loading: boolean;
  error?: Error;
  portfolioId?: string;
  cancelOpenTransaction: typeof cancelTransaction;
  getTransactions: typeof getTransactions;
}

interface TransactionState {
  collapsed: boolean;
  fetched: boolean;
}

export class OpenTransactions extends React.Component<
  TransactionProps,
  TransactionState
> {
  constructor(props: TransactionProps) {
    super(props);
    this.state = { collapsed: true, fetched: false };
  }

  toggleExpanded = () => {
    this.setState({ collapsed: !this.state.collapsed });
    if (!this.state.fetched && this.props.portfolioId) {
      this.props.getTransactions(this.props.portfolioId);
      this.setState({ fetched: true });
    }
  };

  renderToggleButton = () => {
    const iconName = this.state.collapsed ? 'arrowDown' : 'arrowUp';
    return (
      <View style={{ flex: 1, flexDirection: 'row' }}>
        <Icon iconName={'opensign'} iconHeight={24} iconWidth={24} />
        <View style={{ flex: 1, flexDirection: 'row' }}>
          <Text style={portfolioStyles.titleText}>
            {t('OpenTransactions.ToggleTitle')}
          </Text>
          {this.props.loading && (
            <ActivityIndicator
              size={verticalScale(24)}
              style={{ marginLeft: scale(18) }}
            />
          )}
        </View>
        <Icon iconName={iconName} iconHeight={24} iconWidth={24} />
      </View>
    );
  };

  renderContent = () => {
    const { error, transactions, portfolioId } = this.props;
    if (error) {
      return (
        <Text style={portfolioStyles.noActionsText}>
          Error! {error.message}
        </Text>
      );
    }
    if (transactions.length < 1 || portfolioId === undefined) {
      return (
        <Text style={portfolioStyles.noActionsText}>
          {t('OpenTransactions.NoTransactions')}
        </Text>
      );
    }

    return transactions.map((item, i) => (
      <View key={i} style={portfolioStyles.transaction}>
        <Text style={portfolioStyles.symbol}>{item.symbol}</Text>
        <View style={portfolioStyles.portfolioinfo}>
          <View style={{ flex: 1 }}>
            <Text style={portfolioStyles.basicText}>
              {t('OpenTransactions.Amount')}
            </Text>
            <Text style={portfolioStyles.value}>{item.amount}</Text>
          </View>
          <View style={{ flex: 1.5 }}>
            <Text style={portfolioStyles.basicText}>
              {t('OpenTransactions.Total')}
            </Text>
            <Text style={portfolioStyles.value}>{item.price} â‚¬</Text>
          </View>
          <TouchableOpacity
            onPress={() =>
              Alert.alert(
                t('OpenTransactions.AlertTitle'),
                t('OpenTransactions.ConfTxtP1') +
                  item.symbol +
                  t('OpenTransactions.ConfTxtP2'),
                [
                  {
                    text: t('OpenTransactions.N'),
                    style: 'cancel',
                  },
                  {
                    text: t('OpenTransactions.Y'),
                    onPress: () =>
                      this.props.cancelOpenTransaction(portfolioId, item.uid),
                  },
                ]
              )
            }
          >
            <Text style={portfolioStyles.cancelAction}>
              {t('OpenTransactions.CancelAction')}
            </Text>
          </TouchableOpacity>
        </View>
      </View>
    ));
  };

  render() {
    return (
      <View>
        <TouchableOpacity onPress={this.toggleExpanded}>
          {this.renderToggleButton()}
        </TouchableOpacity>
        {!this.props.loading && (
          <Collapsible collapsed={this.state.collapsed}>
            {this.renderContent()}
          </Collapsible>
        )}
      </View>
    );
  }
}
