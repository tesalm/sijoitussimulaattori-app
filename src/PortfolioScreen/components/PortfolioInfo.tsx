import React from 'react';
import { ActivityIndicator, Text, View } from 'react-native';

import { t } from '../../assets/i18n';
import { Stock } from '../../MarketScreen/reducer';
import { formatRevenue, revenueColor } from '../../util/stock';
import { Portfolio, PortfolioStock } from '../reducers';
import { portfolioStyles } from '../styles';

export interface PortfolioInfoProps {
  portfolio?: Portfolio;
  loading?: boolean;
  error?: Error;
  stocks?: Stock[];
}

export const PortfolioInfo = (props: PortfolioInfoProps): JSX.Element => {
  if (props.loading) {
    return (
      <View style={portfolioStyles.loading}>
        <ActivityIndicator size="large" />
      </View>
    );
  } else {
    if (props.portfolio === undefined) {
      let errorMessage;
      if (props.error) {
        errorMessage = props.error.message + ' ';
      }
      // TODO: Muokkaa error-teksti käyttäjälle.
      return <Text>Error! {errorMessage} </Text>;
    }
  }

  return (
    <View>
      <View style={portfolioStyles.portfolioinfo}>
        <Text style={portfolioStyles.titleStyle}>
          {t('PortfolioPage.Title')}
        </Text>
      </View>
      <View style={portfolioStyles.portfolioinfo}>
        <View style={portfolioStyles.portfolioInfoSmallerComp}>
          <Text style={portfolioStyles.valueHeader}>
            {t('PortfolioPage.MarketValue')}
          </Text>
          <Text style={portfolioStyles.value}>
            {props.portfolio.totalMarketValue + ' $'}
          </Text>
          <Text style={portfolioStyles.valueHeader}>
            {t('PortfolioPage.Cash')}
          </Text>
          <Text style={portfolioStyles.value}>
            {props.portfolio.balance + ' $'}{' '}
          </Text>
        </View>
        <View style={portfolioStyles.portfolioInfoSmallerComp}>
          <Text style={portfolioStyles.valueHeaderMiddle}>
            {t('PortfolioPage.TotalValue')}
          </Text>
          <Text style={portfolioStyles.valueMiddle}>
            {props.portfolio.balance + props.portfolio.totalMarketValue + ' $'}
          </Text>
        </View>

        <View style={portfolioStyles.portfolioInfoSmallerComp}>
          <Text style={portfolioStyles.valueHeaderRightSide}>
            {t('PortfolioPage.Revenue')}
          </Text>
          <Text style={revenueColor(props.portfolio.revenue)}>
            {formatRevenue(props.portfolio.revenue)}
          </Text>
        </View>
      </View>
    </View>
  );
};

const calculatePortfolioValue = (
  portfolioStocks: PortfolioStock[],
  stocks?: Stock[]
) => {
  //find right stock from the array.
  let portoflioValue = 0;
  portfolioStocks.forEach((stock) => {
    if (stocks) {
      const rightStock = stocks.find(
        (findStock) => findStock.symbol == stock.symbol
      );
      if (rightStock == undefined) {
        return 0;
      }
      if (rightStock.stockInfo.intraday === undefined) {
        return 0;
      }
      portoflioValue =
        portoflioValue +
        rightStock.stockInfo.intraday.intradayQuote[0].close * stock.amount;
    }
  });
  return portoflioValue;
};

export default PortfolioInfo;
