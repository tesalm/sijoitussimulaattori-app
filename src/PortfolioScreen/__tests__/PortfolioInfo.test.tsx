import { shallow } from 'enzyme';
import * as React from 'react';
import configureStore from 'redux-mock-store';

import { PortfolioInfo, PortfolioInfoProps } from '../components/PortfolioInfo';

describe('portfolioInfo tests', () => {
  // Mock for navigation.
  const navigationMock: any = {};
  const middlewares: any = [];
  const mockStore = configureStore(middlewares);

  const defaultPortfolioProps: PortfolioInfoProps = {
    portfolio: undefined,
    loading: undefined,
    error: undefined,
    stocks: undefined,
  };

  const successPortfolioProps: PortfolioInfoProps = {
    portfolio: {
      uid: 'Portfolio1',
      name: 'Portfolio1',
      balance: 500,
      totalRevenue: -0.02,
      totalMarketValue: 5000,
      lastDayRevenue: 0,
      ownerId: 'Joooni',
      stocks: [
        {
          uid: 'AMZN',
          amount: 245,
          avgPrice: 153.3,
          totalRevenue: 0.54,
          totalMarketValue: 2084,
          lastDayRevenue: 0.52,
        },
        {
          uid: 'AAPL',
          amount: 340,
          avgPrice: 132,
          totalRevenue: 0.54,
          totalMarketValue: 2084,
          lastDayRevenue: 0.52,
        },
      ],
    },
    stocks: [
      {
        symbol: 'AAPL',
        name: 'Apple',
        high: 375.53,
        low: 354.38,
        revenue: 3894.34,
        close: 293.42,
        currency: 'USD',
        stockInfo: {
          stockMetadata: {
            symbol: 'AAPL',
            name: 'Apple Inc.',
            type: 'string',
            region: 'string',
            marketOpen: '2018-12-12 11:20:00',
            marketClose: '2018-12-12 11:20:00',
            timeZone: 'US/Eastern',
            currency: 'USD',
            fetchTime: new Date('2018-12-17T01:24:00.000Z'),
          },
          intraday: {
            intradayQuote: [
              {
                symbol: 'AAPL',
                date: '2018-12-12T00:20:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 165.43,
                close: 168.63,
                volume: 12279994,
              },
              {
                symbol: 'AAPL',
                date: '2018-12-12T00:10:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 165.43,
                close: 168.63,
                volume: 12279994,
              },
            ],
            fetchTime: new Date('2018-12-17T01:24:00.000Z'),
          },
          historyData: {
            historyDataQuote: [
              {
                symbol: 'AAPL',
                date: '2018-12-15T00:00:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 165.43,
                close: 162.63,
                volume: 12279994,
              },
              {
                symbol: 'AAPL',
                date: '2018-12-14T00:00:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 161.43,
                close: 168.63,
                volume: 12279994,
              },
            ],
            fetchTime: new Date('2018-12-17T01:24:00.000Z'),
          },
          metaLoading: false,
          intraLoading: false,
          historyLoading: false,
          refreshing: false,
          metaError: undefined,
          intraError: undefined,
          historyError: undefined,
        },
      },
    ],
    error: undefined,
    loading: false,
  };

  it('PortfolioInfo renders correctly', async () => {
    const wrapper = shallow(
      <PortfolioInfo {...defaultPortfolioProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('PortfolioInfo renders correctly with portfolio and stocks', async () => {
    const wrapper = shallow(
      <PortfolioInfo {...successPortfolioProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
