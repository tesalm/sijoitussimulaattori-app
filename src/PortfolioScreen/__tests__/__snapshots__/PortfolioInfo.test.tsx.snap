// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portfolioInfo tests PortfolioInfo renders correctly 1`] = `
<Text>
  Error! 
   
</Text>
`;

exports[`portfolioInfo tests PortfolioInfo renders correctly with portfolio and stocks 1`] = `
<View>
  <View
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#004D40",
          "fontSize": 34,
          "fontWeight": "bold",
          "marginBottom": 32,
        }
      }
    >
      Portfolio1
    </Text>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 0.8,
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 24,
            "marginBottom": 0,
            "textAlign": "justify",
          }
        }
      >
        Market value
      </Text>
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 27.46470588235294,
            "fontWeight": "bold",
            "marginBottom": 15.694117647058823,
            "marginTop": -3.9235294117647057,
            "textAlign": "justify",
          }
        }
      >
        4500 $
      </Text>
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 24,
            "marginBottom": 0,
            "textAlign": "justify",
          }
        }
      >
        Cash
      </Text>
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 27.46470588235294,
            "fontWeight": "bold",
            "marginBottom": 15.694117647058823,
            "marginTop": -3.9235294117647057,
            "textAlign": "justify",
          }
        }
      >
        500 $
         
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 0.4,
          "flexDirection": "column",
          "marginTop": 12.857142857142858,
          "textAlign": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 85.71428571428571,
          }
        }
      >
        }
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
          "marginTop": 42.857142857142854,
          "textAlign": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 24,
            "marginBottom": 0,
            "textAlign": "justify",
          }
        }
      >
        Total Value
      </Text>
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 27.46470588235294,
            "fontWeight": "bold",
            "marginBottom": 15.694117647058823,
            "marginTop": -3.9235294117647057,
            "textAlign": "justify",
          }
        }
      >
        5000 $
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 0.8,
          "flexDirection": "column",
          "marginTop": 42.857142857142854,
          "textAlign": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#004D40",
            "fontSize": 24,
            "marginBottom": 0,
            "textAlign": "justify",
          }
        }
      >
        Revenue today
      </Text>
      <Text
        style={
          Object {
            "color": "#54c242",
            "fontSize": 36,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        +0.00 %
      </Text>
    </View>
  </View>
</View>
`;
